From b0aa9d7dfe80284b1d23166f4cff24a370ad0929 Mon Sep 17 00:00:00 2001
From: Niall Pemberton <niallp@apache.org>
Date: Thu, 28 Jan 2010 17:05:54 +0000
Subject: [PATCH] Port LANG-552 to 2.x branch - StringUtils.replaceEach(String,
 String[], String[]) no longer NPEs when null appears in the last String[]

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/lang/branches/LANG_2_X@904158 13f79535-47bb-0310-9956-ffa450edef68
---
 src/main/java/org/apache/commons/lang/StringUtils.java     | 3 +++
 src/test/java/org/apache/commons/lang/StringUtilsTest.java | 4 ++++
 2 files changed, 7 insertions(+)

diff --git a/src/main/java/org/apache/commons/lang/StringUtils.java b/src/main/java/org/apache/commons/lang/StringUtils.java
index 4c7dccf92db..554c984ec14 100644
--- a/src/main/java/org/apache/commons/lang/StringUtils.java
+++ b/src/main/java/org/apache/commons/lang/StringUtils.java
@@ -3727,6 +3727,9 @@ private static String replaceEach(String text, String[] searchList, String[] rep
 
         // count the replacement text elements that are larger than their corresponding text being replaced
         for (int i = 0; i < searchList.length; i++) {
+            if (searchList[i] == null || replacementList[i] == null) {
+                continue;
+            }
             int greater = replacementList[i].length() - searchList[i].length();
             if (greater > 0) {
                 increase += 3 * greater; // assume 3 matches
diff --git a/src/test/java/org/apache/commons/lang/StringUtilsTest.java b/src/test/java/org/apache/commons/lang/StringUtilsTest.java
index c3a35ff954e..939fb0e57e7 100644
--- a/src/test/java/org/apache/commons/lang/StringUtilsTest.java
+++ b/src/test/java/org/apache/commons/lang/StringUtilsTest.java
@@ -1071,6 +1071,10 @@ public void testReplace_StringStringArrayStringArray() {
                 "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "N", "O", "P", "Q", 
                 "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "A", "B", "C", "D", "E", "F", "G", 
                 "H", "I", "J", "K", "L", "M", "5", "6", "7", "8", "9", "1", "2", "3", "4"}));
+
+        // Test null safety inside arrays - LANG-552
+        assertEquals(StringUtils.replaceEach("aba", new String[]{"a"}, new String[]{null}),"aba");
+        assertEquals(StringUtils.replaceEach("aba", new String[]{"a", "b"}, new String[]{"c", null}),"cbc");
     }
 
     /**
